intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp

```diff
@@ -414,9 +459,13 @@ std::vector<std::vector<model_token>> Model::generate_tokens(const std::vector<s
   return rets;
 }
 
-model_token Model::post_greedy_search(const float* logits) {
-  model_token id = std::max_element(logits, logits + n_vocab) - logits;
-  return id;
+std::vector<model_token> Model::post_greedy_search(const float* logits) {
+  std::vector<model_token> ids(ctx->batch_size);
+#pragma omp parallel for
+  for (int bs = 0; bs < ctx->batch_size; ++bs) {
+    ids[bs] = std::max_element(logits + bs * n_vocab, logits + (bs + 1) * n_vocab) - (logits + bs * n_vocab);
```

Comment:
I think in most cast batch_size will be a small number, wonder whether the parallel here will be helpful? you can try the following
```suggestion
#pragma omp parallel for num_threads(ctx->batch_size > xxx ? x : 1)
  for (int bs = 0; bs < ctx->batch_size; ++bs) {
    ids[bs] = std::max_element(logits + bs * n_vocab, logits + (bs + 1) * n_vocab) - (logits + bs * n_vocab);
```

Reply:
the `n_vocab` is different between different models (`GPT-J` ~ 5W, `CHATGLM` ~13W). It may be hard to define `xxx`?However, I will try the `num_threads=bs` and `num_threads=1` in this example.

Reply:
50K is a large number, then you can try more threads like
```suggestion
static n_vocab_segment = 1024;
int num_segments = (n_vocab +n_vocab_segment -1 )/ n_vocab_segment;
std::vector<model_token> temp(ctx->batch_size * num_segments);
#pragma omp parallel for collapse(2)
  for (int bs = 0; bs < ctx->batch_size; ++bs) {
    for (int vocab = 0; vocab <n_vocab + 1023; vocab += 1024 ) {
      temp[bs * num_segments + vocab / 1024] = std::max_element(logits + bs * n_vocab + vocab, vocab > n_vocab ? logits + bs * n_vocab + n_vocab : logits + bs * n_vocab + vocab) - (logits + bs * n_vocab);
      }
    ids[bs] = std::max_element(temp + bs *num_segments  , temp + (bs +1)*num_segments) - (logits + bs * n_vocab);
```

Reply:
Yes, thanks for your suggestion. It does help a lot. updated in https://github.com/intel/intel-extension-for-transformers/pull/700/commits/19a5b8872950eb5f87c9c0a8d4efd8db5323c33f.

here is the test reference(icx,1ins /56 core, 128 new_tokens, batch=4, q40)
| omp threads |  time (per_token,ms) |
| -----       | ----       |
|  4          |  0.248     |
|  3           |  0.406    |
| 2            |  0.548    |
| 1            |  0.783  |
| collapse | 0.047 |